File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if height is None or len(height) == 0:
            return 0
        
        left = 0
        right = len(height) - 1
        lw = 0
        rw = 0
        result = 0

        while left <= right:
            if lw <= rw:
                if lw > height[left]:
                    result += lw - height[left]
                    left += 1
                else:
                    lw = height[left]
                    left += 1
            else:
                if rw > height[right]:
                    result += rw - height[right]
                    right -= 1
                else:
                    rw = height[right]
                    right -= 1
        
        return result

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        if(height == null || height.length == 0){
            return 0;
        }

        int left = 0;
        int right = height.length - 1;
        int lw = 0;
        int rw = 0;
        int result = 0;

        while(left <= right){
            if(lw <= rw){
                if(lw > height[left]){
                    result += lw - height[left];
                    left += 1;
                }
                else{
                    lw = height[left];
                    left++;
                }
            }

            else{
                if(rw > height[right]){
                    result += rw - height[right];
                    right -= 1;
                }
                else{
                    rw = height[right];
                    right -= 1;
                }
            }
        }

        return result;
    }
}


// 2 pointers
//Time Complexity: O(n)
//Space Complexity: O(1)